### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by moonlimb**

### CODE REVIEW

File: ltr-end-to-end.sh [hunk 1]
```diff
@@ -34,6 +34,8 @@ done
 shift $((OPTIND -1))
 
 
+
+
```
Comment (moonlimb) [inline ltr-end-to-end.sh]:
```suggestion
```

File: week2/utilities/build_ltr.py [hunk 2]
```diff
@@ -19,6 +19,9 @@
 from opensearchpy import OpenSearch
 
 
+# Ignore pandas Index64 deprecation warning 
+import warnings
+warnings.simplefilter(action='ignore', category=FutureWarning)
```
Comment (itanev) [inline week2/utilities/build_ltr.py]: It's usually a bad idea to ignore warnings. 😉

File: week2/utilities/click_models.py [hunk 3]
```diff
@@ -28,11 +34,16 @@ def apply_click_model(data_frame, click_model_type="binary", downsample=True):
...
+        # print("IMPLEMENT ME: apply_click_model(): downsampling")
+        if downsample:
+            data_frame = down_sample_buckets(data_frame)
     return data_frame
 
 # https://stackoverflow.com/questions/55119651/downsampling-for-more-than-2-classes
 def down_sample_buckets(data_frame):
+    """
+    ASK: what does this function do?
+    """
```
Comment (itanev) [inline week2/utilities/click_models.py:46]: I'm not 100% sure myself, but in general, it seems it's retrieving the grade groups with the smallest size. I can't say I get the purpose though.

File: week2/utilities/data_prepper.py [hunk 4]
```diff
@@ -233,22 +233,36 @@ def __log_ltr_query_features(self, query_id, key, query_doc_ids, click_prior_que
                                                 self.ltr_store_name,
                                                 size=len(query_doc_ids), terms_field=terms_field)
         # IMPLEMENT_START --
-        print("IMPLEMENT ME: __log_ltr_query_features: Extract log features out of the LTR:EXT response and place in a data frame")
+
+        # Run the query just like any other search
+        response = self.opensearch.search(body=log_query, index=self.index_name)
```
Comment (itanev) [inline week2/utilities/data_prepper.py]: What if this throws an exception?

File: week2/utilities/ltr_utils.py [hunk 5]
```diff
@@ -50,8 +76,57 @@ def create_sltr_hand_tuned_query(user_query, query_obj, click_prior_query, ltr_m
...
+
+    # Create our SLTR query
+    query_obj = {
+        'size': size,
+        'query': {
+            'bool': {
+                "filter": [  # use a filter so that we don't actually score anything
+                    {
+                        "terms": {
+                            "_id": doc_ids
```
Comment (itanev) [inline week2/utilities/ltr_utils.py]: Instead of the `_id` you should use the `terms_field`

**Review by itanev — CHANGES_REQUESTED** [review]
Looks good overall. Several small stuff I noticed.

File: week2/utilities/data_prepper.py (hunk 4 above)
Comment (moonlimb) [reply week2/utilities/data_prepper.py]: good point, it'd be nice to wrap this with try / catch :)

File: week2/utilities/ltr_utils.py (hunk 5 above)
Comment (moonlimb) [reply week2/utilities/ltr_utils.py]: thank you!

**Pull request merged**